UnitySDK/Assets/ML-Agents/Scripts/Brain.cs

```diff
@@ -18,21 +19,32 @@ public abstract class Brain : ScriptableObject
     {
         [SerializeField] public BrainParameters brainParameters;
 
-        protected Dictionary<Agent, AgentInfo> m_AgentInfos =
-            new Dictionary<Agent, AgentInfo>(1024);
+        protected List<Agent> m_Agents = new List<Agent>(1024);
 
-        [System.NonSerialized]
+        protected ICommunicator m_Communicator;
+
+        [NonSerialized]
         private bool m_IsInitialized;
 
         /// <summary>
-        /// Adds the data of an agent to the current batch so it will be processed in DecideAction.
+        /// Sets the Batcher of the Brain. The brain will call the communicator at every step and give
+        /// it the agent's data using PutObservations at each DecideAction call.
+        /// </summary>
+        /// <param name="communicator"> The Batcher the brain will use for the current session</param>
+        public void SetCommunicator(ICommunicator communicator)
+        {
+            m_Communicator = communicator;
+            LazyInitialize();
+        }
+
+        /// <summary>
+        /// Updates the data of an agent to the current batch so it will be processed in DecideAction.
         /// </summary>
         /// <param name="agent"></param>
-        /// <param name="info"></param>
-        public void SendState(Agent agent, AgentInfo info)
+        public void Update(Agent agent)
         {
             LazyInitialize();
-            m_AgentInfos[agent] = info;
+            m_Agents.Add(agent);
```

Comment:
(summarizing slack thread)
This feels odd because we always append to the agent list here, and rely on the implementations to clear it. It's slightly worse than before, because if something prevents the clearing (i.e. an exception due to bad configuration), the list grows unbounded now, instead of overwriting when it was a dictionary.

Not sure if we should leave as is, convert to a Set, or change how the clearing happens.

@vincentpierre Ideas?

Reply:
changed here 588466436341384aff58b55ff79cd811719f522f